### DISCUSSION TITLE
Fixed grid header row/column

### DISCUSSION THREAD

**Discussion by PPakalns** [DioxusLabs/taffy#770] [Q&A] [answered]
Similar to this codepen example. https://codepen.io/neoky/pen/mGpaKN

Is it possible to make sticky column, header in `display: Grid, overflow: Scroll` elements.

It seems that it is dependent on last element that is scroll-able and what scroll offset it has.
Technically it looks like i could store "stickiness" flag as metadata and I just shouldn't apply scroll offset to element and it would be drawn correctly.  If this is all that is required, could taffy add `position: sticky` and add a flag to layout if it should be sticky?

Is there another way to achieve sticky header column/row in grid?

**Comment by PPakalns** [comment]
Got behavior similar to position: sticky working in egui_taffy by manually storing `sticky` flag for nodes.

https://github.com/user-attachments/assets/07546146-7a90-422b-b291-99b758fd7704

**Reply by nicoburns** [reply, marked as answer]
> Technically it looks like i could store "stickiness" flag as metadata and I just shouldn't apply scroll offset to element and it would be drawn correctly. If this is all that is required, could taffy add position: sticky and add a flag to layout if it should be sticky?

So I think the "correct" computation for `position: sticky` is a little more complicated than just not applying the scroll offset:
1. It should *conditionally* not apply the scroll offset once the item would otherwise scroll out of view
2. The position at which the "sticky" position takes effect should be inset by the `top/left/bottom/right` styles.
3. There might need to be some kind of conflict resolution for over-constrained cases (e.g. container is smaller than sticky node or sticky node has both `bottom` and `top` set)

But, as you have demonstrated:

- (1) doesn't apply if the initial (static) position of the node is 0
- (2) doesn't apply if the sticky position is 0
- (3) doesn't apply if the container is big enough and only one property is each axis is set

---

I don't think it makes sense to have `position: sticky` as part of the main layout phase in Taffy, because `position: sticky` is scroll position dependent, and you don't want to have to recompute layout every time the scroll position changes. But:

- There could be a helper which computes the position given inputs (which would be called manually)
- It could potentially be incorporated as part of a (new) phase that could:
   - Compute `position: sticky`
   - Apply `transform`
   
I'd probably want to start with the former approach and move towards the latter once I/we are more sure how it needs to work.

**Reply by PPakalns** [reply]
Thank you for the detailed explanation.

If better support for scrollable elements and `position: sticky` is added to taffy in future, helper functions would be very useful to correctly handle scrolling and `position: sticky`.

At least for basic case simple sticky flag for now works good enough in `egui_taffy`.  Thank you all for this great project.
